Bidirectional version of Dijkstra's shortest path

d-michail opened the pull request:
This is an implementation of a bidirectional version of Dijkstra's shortest path algorithm (closes issue #135).

The algorithm maintains two frontiers and alternates between them until they meet.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java:
```diff
@@ -0,0 +1,490 @@
+                source,
+                target,
+                Arrays.asList(source),
+                new ArrayList<>(),
```
Consider: Collections.emptyList()

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java:
```diff
@@ -0,0 +1,490 @@
+                forwardFrontier.graph,
+                source,
+                target,
+                Arrays.asList(source),
```
Consider Collections.singletonList(source)

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java:
```diff
@@ -0,0 +1,490 @@
+        V target,
+        double radius)
+    {
+        this.forwardFrontier = new SearchFrontier<V, E>(graph);
```
Consider replacing <V,E> by <> (also elsewhere, no longer needed in java 8)

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java:
```diff
@@ -0,0 +1,490 @@
+    }
+
+    /**
+     * @see CrossComponentIterator.Specifics#edgesOf(Object)
```
Not sure why you reference to the CrossComponentIterator here? I think that the qualifier "import org.jgrapht.traverse.CrossComponentIterator;" is also missing.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java:
```diff
@@ -0,0 +1,490 @@
+    E e;
+    V v;
+
+    QueueEntry()
```
empty constructor may be omitted.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java:
```diff
@@ -0,0 +1,490 @@
+class BidirectionalDijkstraShortestPathDetails<V, E>
+{
+    private final SearchFrontier<V, E> forwardFrontier;
+    private final SearchFrontier<V, E> reversedFrontier;
```
Consider backwardFrontier (as opposite of forwardFrontier)

jkinable commented:
Looks good to me. Several people have asked for a Bidirectional shortest path in the past, so this is a nice addition. I've included some minor remarks. Let's wait to see whether @jsichi has any additional remarks.

d-michail commented:
Thanks for the comments. I will fix them as soon as possible. 

Regarding the class structure, the public class BidirectionalDijkstraShortestPath has the same structure as DijkstraShortestPath. If you see the test cases, I have an example where I execute both and test whether the result is the same.

Class BidirectionalDijkstraShortestPathDetails is not exported outside the file. I wrote it like that in order to allow the garbage collector to regain used memory even if the caller keeps a BidirectionalDijkstraShortestPath instance around for a long time.

d-michail commented:
Fixed the minor issues. 

I also added a static findPathBetween(V a, V b) method in the BidirectionalDijkstraShortestPath class. Now the interface is exactly the same as in DijkstraShortestPath.

jkinable commented:
Looks good! Btw, nice job on the extensive unit testing and proper documentation.

In the future it would be nice if findPathBetween in both Dijkstra classes, would actually return a GraphPath as the name suggests (instead of a vertex path). That way the user has instant access to a full description of the path in terms of edges, vertices and cost. That would also work for multigraphs. For now I suggest we keep it like it is.

jsichi commented:
lgtm.  @jkinable do you want to go ahead and merge the pull request?

(As a followup, perhaps we should add a new interface to org.jgrapht.alg.interfaces to make the commonality explicit?)

Regarding the findPathBetween convenience method, we could deprecate it and add the improved version to java.org.jgrapht.Graphs (with an extra parameter controlling whether to use bidirectional search).

jkinable commented:
Thanks @d-michail , nice contribution!

We should indeed add a new interface to org.jgrapht.alg.interfaces which covers single source-target paths.
